storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/dataaccess/V3RocksDbDao.java

```diff
@@ -96,6 +96,25 @@ public V3RocksDbDao(final RocksDbAccessor db) {
     return getFinalizedBlock(blockRoot).map(SignedBeaconBlock::getSlot);
   }
 
+  @Override
+  public Optional<UnsignedLong> getSlotForFinalizedStateRoot(final Bytes32 stateRoot) {
+    return db.get(V3Schema.SLOTS_BY_FINALIZED_STATE_ROOT, stateRoot);
+  }
+
+  @Override
+  public Optional<SlotAndBlockRoot> getSlotAndBlockRootForFinalizedStateRoot(
+      final Bytes32 stateRoot) {
+    Optional<UnsignedLong> maybeSlot = db.get(V3Schema.SLOTS_BY_FINALIZED_STATE_ROOT, stateRoot);
+    if (maybeSlot.isPresent()) {
+      Optional<SignedBeaconBlock> maybeRoot = getFinalizedBlockAtSlot(maybeSlot.get());
+      if (maybeRoot.isPresent()) {
+        return Optional.of(
+            new SlotAndBlockRoot(maybeSlot.get(), maybeRoot.get().getMessage().hash_tree_root()));
+      }
+    }
+    return Optional.empty();
```

Comment:
Probably better to do this with a `.flatMap` rather than the if style. I think it becomes:
```suggestion
    return maybeSlot.flatMap(slot -> 
        getFinalizedBlockAtSlot(slot).map(block -> new SlotAndBlockRoot(slot, block.getRoot())));
```